### ISSUE TITLE
Kubevirt doesn't report detailed VMI metrics

### ISSUE THREAD

**Issue by fromanirh** [kubevirt/kubevirt#1594]
/kind enhancement

**What happened**:
In order to monitor, troubleshoot and tune VMs, we would like to expose detailed metrics about how the VMI instance is behaving. We are concerned about the infrastructure here, so the infrastructure processes running inside the virt-launcher pod.
For example, we want to assess the overhead of the aformenentioned infrastructure, allowing to have answers to queries like
- how much memory is libvirtd taking?
- is that memory growing?
- how much memory QEMU takes?
- is libvirtd leaking threads?

In addition, for the same purposes, we would like to have a way to expose the metrics reported to libvirt (see "virsh domstats").

At the moment, there is no way to gather this information unless we do this manual work.
We would like to have an automated flow to do that.

This enhancement should integrate smoothly with prometheus operator, and should build on top of PR https://github.com/kubevirt/kubevirt/pull/1515

**What you expected to happen**:
Have all the metrics in prometheus.
Additionally, it would be nice to have a simple way to add more metrics should they be needed.
"Simple" is something in the same ballpark of "editing a configmap"

**How to reproduce it (as minimally and precisely as possible)**:
N/A

**Anything else we need to know?**:
We thought of two possible approach about to fix that, I'm going to illustrate them in the comments

**Environment**:
N/A (any)

**Comment by fromanirh** [comment]
In general, we need to extract infos from processes running inside the virt-launcher compute container.
We thought about two possible way to implement this enhancement.

1. enhance the virt-launcher image to collect the metrics. We can do that using an image like this:
https://github.com/fromanirh/libvirt-collectd
This image is a drop-in replacement over `kubevirt/libvirtd`, using it is trivial: https://github.com/fromanirh/kubevirt/commit/007112d9ef3a2fdd654221cc2666d3edbdbb01fc

2. we need a way to exfiltrate the stats we now collect. Since the network configuration of the pod is passed to the VM, we cannot just expose the collectd instance running inside the pod.

3. to fix that we could: add a shared temporary volume on which each pod will dump its stats, and make node_exporter collecting and reporting them using the `textfile` collector.
This requires another trivial patch to virt-controller (to create the temporary directory) and a patch to prometheus operator to consume this data.

The drawback of this flow is that it is complex, has many moving parts, makes the virt-launcher pod quite crowded (virt-launcher + libvirtd + qemu + collectd + maybe other helpers), adds quite some delay in the reporting chain because every item has to poll the next - so the intervals add up, and requires patches to 3rd party components - prometheus operator and collectd.

**Comment by fromanirh** [comment]
Another possible option is to add a per-host (no longer per-pod) collectd instance to collect all the needed information at once. This will be a daemonset which can be deployed side-by-side to the prometheus operator pods and daemonsets. Highlights of this solution are

1. we will have a new daemonset featuring collectd and a helper program; this will need to scan the *host* /proc to collectd data about libvirtd and qemu behaviour
2. in order to map the process to pods, we will need to connect either to the CRI socket or to kubelet. otherwise we will report unhelpful stats about PIDs, while users (most likely) care about PODs or VM(I)s
(I have POC code to find the POD on which a given pid is running)
3. the new daemonset will expose an additional prometheus endpoint to be queried (this should not be a problem, and it is actually a nice incremental approach)
4. no changes required to virt-launcher, virt-controller, images or prometheus operator
5. it is not clear how to gather the virt metrics (virsh domstats) at the moment, but there is no clear need for them either at this moment

I have code, images and manifests for all the above too.

**Comment by fromanirh** [comment]
> @fromanirh is it even worth bringing in collectd into the picture here?

Collectd is just the path of least resistance, I can't (and won't) claim is the best approach here.

> If we can, I would prefer a simple in-code solution which works like that:
> 
> 1)virt-launcher binary collects metrics,
> 2) virt-launcher exposes metrics directly **or** virt-handler exposes metrics for all virt-launchers

This is totally doable and IMHO a much nicer long-term solution.
I think I can adapt my POC code to integrate it into virt-handler.
Gathering metrics from virt-launchers is also fine, granted we figure out how to collect and expose them since virt-launcher cannot be a metrics endpoint (no network access)

> It looks like we only want to monitor a few processes and additionally ask libvirt for some metrics, which we would convert into prometheus metrics. That should not be too difficult. Thoughts?

I agree with this assessment.

**Comment by rmohr** [comment]
> Gathering metrics from virt-launchers is also fine, granted we figure out how to collect and expose them since virt-launcher cannot be a metrics endpoint (no network access)

You can use our RPC connection betwwen virt-handler and virt-launcher: https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-handler/cmd-client/client.go#L55 to collect the metrics from virt-launchers. If possible, we should also live-query all instances when prometheus scrapes the target.

**Comment by rmohr** [comment]
/cc @davidvossel @vladikr @MarSik

Getting your feedback would be great.

**Comment by MarSik** [comment]
Well the point of collectd here is that it works right now. There is no code to write and no code to maintain. We currently lack all the code that would know how to scrape /proc for example.

Maintaining everything ourselves seems unnecessary to me and I hate the NIH syndrome. The scalability concern is a valid one if that is what you are afraid of, but we will have to check and see what happens with both possible solutions.

**Comment by slintes** [comment]
> There is no code to write and no code to maintain. 
I understand that we need to extend the `procwatch` plugin [0]?

> We currently lack all the code that would know how to scrape /proc for example.
I guess the prometheus client provides most of what we need [1].

[0] https://github.com/fromanirh/procwatch
[1] https://github.com/prometheus/client_golang/blob/master/prometheus/process_collector.go#L76

**Comment by fromanirh** [comment]
> > There is no code to write and no code to maintain.
> 
> I understand that we need to extend the `procwatch` plugin [0]?

Yes. Unfortunately this is something we need to do. We cannot simply use stock collectd plugins.

> > We currently lack all the code that would know how to scrape /proc for example.
> 
> I guess the prometheus client provides most of what we need [1].

It seems so, with two caveats:
1. we may need to add more metrics in the future, extending the prometheus client of course works but takes longer - granted the metrics we need fit in the larger prometheus usecase. May or may not be true
2. the issue is how expose the metrics. We will need to extend the scrape chain like Roman explained.
Of course is doable (and like I said we would likely to get there in the future) but requires work too

**Comment by slintes** [comment]
for adding more metrics you don't need to **extend** the prometheus client, you can just **use** it to expose whatever you want. It's only a few lines of code :) See e.g. [0]

[0] https://github.com/prometheus/client_golang/blob/master/examples/random/main.go

**Comment by davidvossel** [comment]
> /cc @davidvossel @vladikr @MarSik Getting your feedback would be great.

I'm in favor of collecting the metrics in our components and exposing them to prometheus. Like it's been suggested, virt-handler can scrape metrics related to the VMI's pod using the cmd channel. 

From what I see so far, anything else is going to involve us introducing more infrastructure and more custom dev to achieve the visibility we really need.

Generic solutions are great for operations teams who need to cast a wide net across their infrastructure, but we're talking about a very detailed look into a single system (KubeVirt). This means being able to correlate things specific to KubeVirt. For example, these generic solutions aren't going to give us visibility into the virt-launcher, qemu and libvirtd processes specific to a single VMI because that's a correlation specific to KubeVirt.

**Comment by vladikr** [comment]
> /cc @davidvossel @vladikr @MarSik Getting your feedback would be great.

I also like the idea of collecting metrics in our components, but I'm a bit worried about the additional overhead we will introduce to each launcher. Would be great if we could find a way to just implement it all in virt-handler.

**Comment by rmohr** [comment]
> we should also consider an equally important perspective which is
> "pluggable add-on VS native integration"
> The current approach I implemented in #1596 is both
>
> 1. customization of an existing solution and
> 2. pluggable add-on

@fromanirh the proposed solutions

1. makes assumptions about kubevirt internals
2. is very complex
3. and while it re-uses an existing solution, it simply reuses a plugin which solely you use and develop for a system which we don't use right now.

So I don't think that your arguments will hold in this context. The same for the comment from @MarSik:

> Maintaining everything ourselves seems unnecessary to me and I hate the NIH syndrome.

I don't think that anyone would for instance ever try to push generic node-monitoring into KubeVirt, which then definitely would be NIH.

My conclusion: I don't think that it has potential to be included as **the** solution to provide metrics for kubevirt. Especially since we will provide VMI metrics through virt-handler/virt-launcher directly and the few extra proc metrics can easily implemented by querying proc in the container if prometheus scrapes the target (not even periodic things needed). So we would only need what's in https://github.com/fromanirh/procwatch/blob/master/procnotify/procnotify.go#L132 after we simply loaded all processes.

Having that said, as discussed offline, you can of course experiment with your solution and use it since it touches nothing in KubeVirt.

**Comment by davidvossel** [comment]
>> /cc @davidvossel @vladikr @MarSik Getting your feedback would be great.
Generic solutions are great for operations teams who need to cast a wide net across their infrastructure, but we're talking about a very detailed look into a single system (KubeVirt). This means being able to correlate things specific to KubeVirt. For example, these generic solutions aren't going to give us visibility into the virt-launcher, qemu and libvirtd processes specific to a single VMI because that's a correlation specific to KubeVirt.

>I understand the sentiment (and I also like that the approach, see previous comments), but we totally can have that visibility - actually, I have POC code about that.

from my previous comment. 

**"From what I see so far, anything else is going to involve us introducing more infrastructure and more custom dev to achieve the visibility we really need."**

yep, I understand we can gain that kind of visibility using something like what you've done in the POC, however it's unclear to me what we're gaining at that point (besides that you've already done some of the work). My concern is we'd be further investing in a more complex solution that we'll later have to abandon. 

> Since we are in the learning stage having the collection as much as possible self-contained and pluggable is a big plus (zero changes needed to kubevirt core :)). Once things we stabilize, we can and should move through the native, in-code solution.

I think it's worth experimenting and leveraging what you already have now to learn/track KubeVirt performance.

However, I don't think we need to wait on anything to stabilize before we start building the native in-code solution. If we're in agreement about our components exposing performance metrics as being the long term solution, we can start that now.

**Comment by fromanirh** [comment]
Thanks everyone for the comments, the suggestions and the thought-provoking discussion.
It is worth to mention that we also reached the point of diminishing returns for collectd: it is starting to prove itself unfit for this very specific task, so we need to start looking for a replacement.

At this point, the in-code solution looks even more interesting.

However, even at this point collectd provided a benefit which I find important: it is trivial to update/replace it while we are experimenting with metrics and explore them.
This is something that I would like very much to preserve, also considering the past experience with oVirt.

I also think this property would not conflict with the suggested in-code solution, I believe we can have all the code ready in nice kubevirt packages, and just invoke it from a different server, instead from, say, virt-handler.

IOW we can expose the metrics using a new simple component called virt-metrics (or any better name) instead from virt-handler.

That new component could be very simple - just trivial wiring, with all the interesting logic in the packages.
Once the exploration stage is done we can happily drop later.
(I for myself believe that keeping the metrics frontend separated is better also once the metrics set is stable, in the long term, but let's save this discussion for another day).

**Comment by fabiand** [comment]
Excuse me if I repeat something, but to summarize the current state:

1. Now: We keep experimenting with collectd outside of the core components
- eventually in a DS.
2. Soon: We aim at adding metrics/monitoring _capabilities_ to the
individual core components (launcher, handler, …)
3. Then: A separate component could use these the capabilities to gather,
then export the metrics using prometheus endpoints

This sounds good to me.
To me in 1. it will be interesting to see what metrics we want to export. I
have some candidates in mind, but the goal to me would be to understand
what is really required by the UI or other tooling.

For 1. and 3. - Can we somehow make sure that a transition between the
implementations is painless for users?

**Comment by fromanirh** [comment]
> Can we somehow make sure that a transition between the implementations is painless for users?

The only possible issue that I see here is the name of the metrics we expose. Once this is stable, the way we collect them should be 100% transparent to users, giving us complete freedom of implementation.

The current implementation we have will most likely still require some work here.

**Comment by fromanirh** [comment]
fixed by https://github.com/kubevirt/kubevirt/pull/1840
